mk/libraries.mk

```diff
@@ -127,7 +127,7 @@ define build-library
     $$($(1)_PATH): $$($(1)_OBJS) | $$(_d)/
 	$(trace-ar) $(AR) crs $$@ $$?
 
-    $(1)_LDFLAGS_USE += $$($(1)_PATH) $$($(1)_LDFLAGS)
+    $(1)_LDFLAGS_USE += -Wl,--whole-archive $$($(1)_PATH) -Wl,--no-whole-archive $$($(1)_LDFLAGS)
```

Comment:
why first one then the other?

Reply:
We only want the whole archive of the local Nix libraries like libnixstore.a. So this enables ```--whole-archive``` for one .a file then immediately disables it. For most dependencies, we only want to include the symbols actually used by the binary. If we enabled ```--whole-archive``` for all libraries, we would end up with tons of unused symbols from libcurl, aws, boehm-gc, etc.

The one case where we need ```--whole-archive``` is for libnixstore.a which has an ```.init``` elf section we need to keep. This comes from the reliance on C++ Global Constructors feature. Some references on this:

 - http://l4u-00.jinr.ru/usoft/WWW/www_debian.org/Documentation/elf/node4.html#SECTION00031000000000000000
- https://github.com/bazelbuild/rules_go/issues/1486
- https://www.bfilipek.com/2018/02/static-vars-static-lib.html